docs/design/pitr-mysql.md

```diff
@@ -0,0 +1,218 @@
+
+# Terminology
+A *physical backup* consists of raw copies of the directories and files that store database contents. This type of backup is suitable for large, important databases that need to be quickly recovered when problems occur. 
+
+A *logical backup* is a logical snapshot of a database, represented by a full logical dump of SQL statements. [This reference](https://dev.mysql.com/doc/refman/8.0/en/backup-types.html) highlights the difference between logical backup and physical backup.
+
+*Incremental backup* is the changes to the database since the last full backup, represented as the binlog in MySQL. Incremental backup can also be called data changes. Incremental backup applies to logical backups.
+
+*Point-in-time Recovery* is the process of making a database's state current up to a given time by restoring the database from a logical backup and applying incremental backup data changes up to the given time.
+
+A *database* is a collection of tables represented as is in MySQL. A *database instance* (instance for short) is the server process that runs the database management system. In MySQL, it’s similar to a `mysqld` process.
+
+
+# Overview
+
+Currently, Bytebase users can set up a backup policy, take full logical backups automatically following the backup policy schedule, and restore the backup to a new database. There are two problems we would like to address in this design.
+
+1. Point-in-time Recovery.
+2. In-place recovery.
+
+We are introducing Point-in-time Recovery (PITR). Consider the scenario that the user made a mistake and dropped a table at 8:00 am, only to find that the last full backup was produced at 2:00 am. If they restore to the previous backup snapshot, new data changes would be lost in the latest 6 hours. With PITR, we can restore the database state to any point between the checkpoints at the granularity of a single database transaction or a timestamp.
+
+We are also introducing In-place Recovery. Without it, users have to change the data source name (DSN) and the database connection string in the software, which is a pain point because it requires
+
+1. Coordination between developers, DBAs, and even network engineers,
+2. Source code modification and software redeployment.
+
+We target database-level recovery with logical backup because the database is the minimal isolation unit within a database instance (an instance of a database server), a natural, logical unit mapping to the business application unit. Logical mistakes usually happen to a single database. Users would like to recover a single database without impacting other databases.
+
+## Requirements
+
+- MySQL 8.0 and 5.7.
+- The database should use the InnoDB storage engine.
+- Enable binlog and choose row-based binlog format.
+- Databases for Point-in-time Recovery are not involved in [XA Transactions](https://dev.mysql.com/doc/refman/8.0/en/xa.html).
```

Comment:
Shall we also add a remaining space check since it will require an extra database size? If that database is too large, we may run out of disk spaces when doing the restore.

Reply:
We can define a watermark, say 1 GB.

if the database is 5GB. Then if it's less than 5GB remaining, it will definitely fail. If there are say 5.5 GB, we give a warning (like our computer tells we are about to run out of space)

Reply:
SG. But it's not easy to get an accurate estimation of the dumped full backup SQL files and binlog files.

Reply:
I guess we can use the size of the existing database as the base. https://stackoverflow.com/questions/1733507/how-to-get-size-of-mysql-database

It's an estimate, so we can just give a warning say there is not much space left, and the restore might fail due to not enough spaces.

Reply:
I tried with sysbench
```
# create a test database with 1m records
sysbench --mysql-user=root --tables=1 --table-size=1000000 oltp_read_write prepare
# check database size
mysql> SELECT table_schema "DB Name",
    ->         ROUND(SUM(data_length + index_length) / 1024 / 1024, 1) "DB Size in MB"
    -> FROM information_schema.tables
    -> GROUP BY table_schema;
+--------------------+---------------+
| DB Name            | DB Size in MB |
+--------------------+---------------+
| mysql              |           2.6 |
...
| sbtest             |         230.2 |
+--------------------+---------------+

# dump database sbtest
mysqldump --databases sbtest --user root > dump
# check dump size
$ ll dump
-rw-r--r--  1 dragonly  staff   191M May  9 10:04 dump
```

It seems like reasonable.